src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs

```diff
@@ -281,272 +194,100 @@ public virtual void Unlock(long position, long length)
-                }
-            }
-
-            return t;
+            return _strategy.ReadAsync(buffer, cancellationToken);
         }
 
         public override void Write(byte[] buffer, int offset, int count)
         {
             ValidateReadWriteArgs(buffer, offset, count);
-            if (_useAsyncIO)
-            {
-                WriteAsyncInternal(new ReadOnlyMemory<byte>(buffer, offset, count), CancellationToken.None).AsTask().GetAwaiter().GetResult();
-            }
-            else
-            {
-                WriteSpan(new ReadOnlySpan<byte>(buffer, offset, count));
-            }
-        }
 
-        public override void Write(ReadOnlySpan<byte> buffer)
-        {
-            if (GetType() == typeof(FileStream) && !_useAsyncIO)
-            {
-                if (_fileHandle.IsClosed)
-                {
-                    throw Error.GetFileNotOpen();
-                }
-                WriteSpan(buffer);
-            }
-            else
-            {
-                // This type is derived from FileStream and/or the stream is in async mode.  If this is a
-                // derived type, it may have overridden Write(byte[], int, int) prior to this Write(ReadOnlySpan<byte>)
-                // overload being introduced.  In that case, this Write(ReadOnlySpan<byte>) overload should use the behavior
-                // of Write(byte[],int,int) overload.  Or if the stream is in async mode, we can't call the
-                // synchronous WriteSpan, so we similarly call the base Write, which will turn delegate to
-                // Write(byte[],int,int), which will do the right thing if we're in async mode.
-                base.Write(buffer);
-            }
+            _strategy.Write(buffer, offset, count);
         }
 
+        public override void Write(ReadOnlySpan<byte> buffer) => _strategy.Write(buffer);
+
         public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
         {
             ValidateBufferArguments(buffer, offset, count);
 
-            if (GetType() != typeof(FileStream))
-            {
-                // If we have been inherited into a subclass, the following implementation could be incorrect
-                // since it does not call through to Write() or WriteAsync() which a subclass might have overridden.
-                // To be safe we will only use this implementation in cases where we know it is safe to do so,
-                // and delegate to our base class (which will call into Write/WriteAsync) when we are not sure.
-                return base.WriteAsync(buffer, offset, count, cancellationToken);
-            }
-
             if (cancellationToken.IsCancellationRequested)
                 return Task.FromCanceled(cancellationToken);
 
-            if (IsClosed)
+            if (_strategy.IsClosed)
                 throw Error.GetFileNotOpen();
```

Comment:
Would it be cheaper if this were in the _strategy.WriteAsync method itself?  Seems like we could potentially avoid one additional virtual dispatch?

Reply:
We could, but here I did this on purpose to keep all the validation logic in `FileStream` and keep all the strategies free of any such checks (mostly to avoid code duplication).